/* Copyright 2017 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

// Protocol buffers for storing metadata for experiments, sessions and metrics.
// TODO(erez): Should these be defined in tensorflow?
syntax = "proto3";

package tensorboard.plugins.hparams_explorer;

message ExperimentMetadata {
  string name = 1;
  string owner = 2;
  string description = 3;
  ConfigType config_type = 4;
}

message SessionMetadata {
  // Contains the actual hyperparameter values for this session
  Config config = 1;

  // An identifier for the session group this session belongs to.
  // All runs in the same session group
  // have the same 'group_id' (and the same 'config') field.
  // Distinct session groups have distinct group_id's.
  string group_id = 2;
}

message ConfigType {
  map<string, ParamType> param_defs = 1;
}

message Config {
  map<string, ParamValue> param_values = 1;
}

enum ParamType {
  FLOAT32 = 0;
  FLOAT64 = 1;
  INT32 = 2;
  INT64 = 3;
  STRING = 4;
  // TODO(erez): Add Tensor types (one for each dtype).
}

message ParamValue {
  oneof value {
    float float32_val = 1;
    double float64_val = 2;
    int32 int32_val = 3;
    int64 int64_val = 4;
    string string_val = 5;
  }
}
